### ISSUE TITLE
[FR] mlflow gc working for s3 artifact-store

### ISSUE THREAD

**Issue by sklingel** [mlflow/mlflow#3953]
Thank you for submitting a feature request. **Before proceeding, please review MLflow's [Issue Policy for feature requests](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md#feature-requests) and the [MLflow Contributing Guide](https://github.com/mlflow/mlflow/blob/master/CONTRIBUTING.rst)**.

**Please fill in this feature request template to ensure a timely and thorough response.**

## Willingness to contribute
The MLflow Community encourages new feature contributions. Would you or another member of your organization be willing to contribute an implementation of this feature (either as an MLflow Plugin or an enhancement to the MLflow code base)?

- [x] Yes. I can contribute this feature independently. (I can try)
- [x] Yes. I would be willing to contribute this feature with guidance from the MLflow community.
- [ ] No. I cannot contribute this feature at this time.

## Proposal Summary
Fill the function 'delete_artifacts'  in 'mlflow/store/s3_artifact_repo.py', instead of throwing an MLFlowExpection

## Motivation
Enable mlflow gc to work with s3 aritfact deletion

### What component(s), interfaces, languages, and integrations does this feature affect?
Components 
- [x] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interfaces
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Languages 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

## Details
I have already a working code snippet:

```
    def delete_artifacts(self, artifact_path=None):

        (bucket, dest_path) = data.parse_s3_uri(self.artifact_uri)

        if artifact_path:
            dest_path = posixpath.join(dest_path, artifact_path)
        s3_client = self._get_s3_client()
        list_objects = s3_client.list_objects(Bucket=bucket, Prefix=dest_path).get('Contents', [])
        for to_delete_obj in list_objects:
            file_path = to_delete_obj.get("Key")
            self._verify_listed_object_contains_artifact_path_prefix(
                listed_object_path=file_path, artifact_path=dest_path
            )
            s3_client.delete_object(Bucket=bucket, Key=file_path)
```

I can create a PR and additional write tests for the function.
However, since would be my first mlflow contribution I guess  I would need some help with the test env.

**Comment by dmatrix** [comment]
@sklingel Thanks for filing this FR. Seems like a reasonable "good-first-issue" to tackle. Yes, it'll help `mlflow gc` for removing growing artifacts in S3 if they are stored there. There was an issue raised in the [slack about this](https://mlflow-users.slack.com/archives/CBFQ95Y2G/p1609924137498800). 

Yes, you'll have to add a test in the [test_s3_artifact_repo.py](https://github.com/mlflow/mlflow/blob/master/tests/store/artifact/test_s3_artifact_repo.py), which checks to see if the uploaded files are deleted. Look at the https://github.com/mlflow/mlflow/blob/master/tests/store/artifact/test_s3_artifact_repo.py#L129 as an example. 

cc: @harupy @dbczumar any advice on writing the delete_artifact_tests?

**Comment by dbczumar** [comment]
> @sklingel Thanks for filing this FR. Seems like a reasonable "good-first-issue" to tackle. Yes, it'll help `mlflow gc` for removing growing artifacts in S3 if they are stored there. There was an issue raised in the [slack about this](https://mlflow-users.slack.com/archives/CBFQ95Y2G/p1609924137498800).
> 
> Yes, you'll have to add a test in the [test_s3_artifact_repo.py](https://github.com/mlflow/mlflow/blob/master/tests/store/artifact/test_s3_artifact_repo.py), which checks to see if the uploaded files are deleted. Look at the https://github.com/mlflow/mlflow/blob/master/tests/store/artifact/test_s3_artifact_repo.py#L129 as an example.
> 
> cc: @harupy @dbczumar any advice on writing the delete_artifact_tests?

Re: testing - https://github.com/mlflow/mlflow/blob/master/tests/store/artifact/test_s3_artifact_repo.py leverages a mock S3 bucket constructed using the `moto` library; this should enable the PR author to assert that deleted files are no longer present in the mock bucket using the same `boto3` client calls that would be used against a real bucket.

**Comment by sklingel** [comment]
Thanks. 
I started working on the test, after some intial setup issue, I found that you have to set
```
export AWS_DEFAULT_REGION=us-east-1
```
It seems that it needs to be set for this specific region (https://stackoverflow.com/questions/49665155/boto3-aws-s3-bucket-creation-error).

Shouldn't that be added to: https://github.com/mlflow/mlflow/blob/master/tests/helper_functions.py#L252?

**Comment by dbczumar** [comment]
> Thanks.
> I started working on the test, after some intial setup issue, I found that you have to set
> 
> ```
> export AWS_DEFAULT_REGION=us-east-1
> ```
> 
> It seems that it needs to be set for this specific region (https://stackoverflow.com/questions/49665155/boto3-aws-s3-bucket-creation-error).
> 
> Shouldn't that be added to: https://github.com/mlflow/mlflow/blob/master/tests/helper_functions.py#L252?

@sklingel Good question! The fact that existing tests pass without this configuration seems to indicate that this environment variable may not be entirely necessary. Is there something different about the test case you're adding that results in this behavior?

**Comment by sklingel** [comment]
@dbczumar all the test in https://github.com/mlflow/mlflow/blob/master/tests/store/artifact/test_s3_artifact_repo.py are **not** passing locally for me, without this environment variable set.

**Comment by sklingel** [comment]
What is the approach for the PR? I create a branch and link the issue in the PR?

**Comment by dbczumar** [comment]
> @dbczumar all the test in https://github.com/mlflow/mlflow/blob/master/tests/store/artifact/test_s3_artifact_repo.py are **not** passing locally for me, without this environment variable set.

Are you using the version of `moto` used by the MLflow test environment? https://github.com/mlflow/mlflow/blob/70f28c11f4687ebc089ee79cb920d813cb26e6ca/dev/small-requirements.txt#L10

> What is the approach for the PR? I create a branch and link the issue in the PR?

Yep! That would be great - I'll give it a review when you're ready!

**Comment by sklingel** [comment]
> Are you using the version of moto used by the MLflow test environment?
Yes I'm using moto==1.3.16

I just created the PR. Happy to see your feedback

**Comment by dmatrix** [comment]
thanks, @dbczumar for the quick feedback and @sklingel for filling the PR.
As always, we appreciate community contributions!

**Comment by harupy** [comment]
addressed by #3958
